audio-device-client/explainer.md

```diff
@@ -0,0 +1,167 @@
+with a audio rendering thread.
+
+For querying constraints, a pattern of `mediaDevices.enumerateDevices()` can
+be used. A device client constructor takes a constraint dictionary and generates
+an instance when the query is acceptable by UA. Otherwise, the promise will be
+rejected.
+
+```js
+/* async scope */
+
+const devices = await navigator.mediaDevices.enumerateDevices();
+
+// An imaginary helper function for selecting device IDs for I/O.
+const audioDeviceIds = getMyAudioDeviceIds(devices);
+
+const clientConstraints = {
+  inputDeviceId: audioDeviceIds.input,
+  outputDeviceId: audioDeviceIds.output,
+  sampleRate: 8000,
+  callbackBufferSize: 512,
+  inputChannelCount: 2,
+  outputChannelCount: 6,
+};
+
+const client =
+    await navigator.mediaDevices.getAudioDeviceClient(clientConstraints);
+await client.addModule('my-client.js');
+await client.start();
+```
+
+Operations in `AudioDeviceClientGlobalScope` is somewhat similar to Audio
+Worklet. A callback function should be defined with JS and WASM so it can be
+invoked by UA periodically.
+
+Note that the user code can trigger the `AudioContext` to render its graph by
+calling `contextCallback` function. After that, the rendered data from the
+context can be processed in the client's global scope.
+
+```js
+
+import Processor from './my-audio-processor.js';
+
+// An imaginary function that creates a storage for multi-channel audio data.
+const contextOutput = generateFloat32Arrays(2, callbackBufferSize);
+
+// Main process function of device client's global scope.
+const process = (input, output, contextCallback) {
+  // |input| will be routed to Web Audio API's |context.source| node, and
+  // the result from context renderer will fill up |contextOutput|.
+  contextCallback(input, contextOutput);
+
+  // Takes the data rendered by AudioContext and perform custom processing to
+  // fill out |output| data.
+  Processor.process(contextOutput, output);
+};
+
+setDeviceCallback(process);
+```
+
+Lastly, an instance of `AudioContext` can be obtained from the device client.
+This is optional and user can choose not to use `AudioContext` at all.
+
+```js
+// A client can instantiate an AudioContext.
+const audioContext = client.getContext();
+
+const oscillator = new OscillatorNode(audioContext);
+oscillator.connect(audioContext.destination);
+oscillator.start();
+...
+```
+
+
+## Alternative Designs
+
+
+### Relaxing the "128" rule for `AudioContext`
+
+This idea was quickly turned down by Audio Working Group because it changes the
+fundamental of Web Audio API implementation. Also this does not address
+problems outside of Web Audio API such as I/O hardware configuration.
+
+
+## Conclusion
+
+The Audio Device Client brings the low-level audio functionality to the web
+platform without unnecessary complexity. It provides the closer access to audio
+hardware with configurable parameters such as sample rate, callback buffer size
```

Comment:
Not sure what "configurable parameters" really means here. I think on Android, the sample rate is pretty much fixed and you can't configure it in any way.

Reply:
ADC sits on top of browser's audio service. It will set up necessary infra (e.g. resampler, FIFO) so it can communicate with the service. ADC does not talk to the hardware directly. It's sort of user-facing interface of UA's audio service.